lineapy/graph_reader/artifact_collection.py

```diff
@@ -0,0 +1,335 @@
+        session_function_return = ", ".join(
+            [
+                coll.return_variables[0]
+                for coll in session_artifacts.artifact_nodecollections
+                if coll.collection_type == NodeCollectionType.ARTIFACT
+            ]
+        )
+
+        SESSION_FUNCTION_TEMPLATE = load_plugin_template(
+            "session_function.jinja"
+        )
+        session_function = SESSION_FUNCTION_TEMPLATE.render(
+            indentation_block=indentation_block,
+            session_function_name=session_function_name,
+            session_function_body=session_function_body,
+            return_dict_name=return_dict_name,
+        )
+
+        # Generate calculation code block for the session
+        # This is to be used in multi-session module
+        session_calculation = (
+            f"{indentation_block}artifacts.update({session_function_name}())"
+        )
+
+        return {
+            "session_imports": session_imports,
+            "artifact_functions": artifact_functions,
+            "session_function": session_function,
+            "session_function_return": session_function_return,
+            "session_calculation": session_calculation,
+        }
+
+    def _compose_module(
+        self,
+        session_artifacts_sorted: List[SessionArtifacts],
+        indentation: int = 4,
+    ) -> str:
+        """
+        Generate a Python module that calculates artifacts
+        in the given artifact collection.
+        """
+        indentation_block = " " * indentation
+
+        # Initiate store for module script components
+        module_dict: Dict[str, List[str]] = {
+            "session_imports": [],
+            "artifact_functions": [],
+            "session_function": [],
+            "session_function_return": [],
+            "session_calculation": [],
+        }
+
+        # Extract module script components by session
+        for session_artifacts in session_artifacts_sorted:
+            session_module_dict = self._extract_session_module(
+                session_artifacts=session_artifacts,
+                indentation=indentation,
+            )
+            module_dict["session_imports"].append(
+                session_module_dict["session_imports"]
+            )
+            module_dict["artifact_functions"].append(
+                session_module_dict["artifact_functions"]
+            )
+            module_dict["session_function"].append(
+                session_module_dict["session_function"]
+            )
+            module_dict["session_function_return"].append(
+                session_module_dict["session_function_return"]
+            )
+            module_dict["session_calculation"].append(
+                session_module_dict["session_calculation"]
+            )
+
+        # Combine components by "type"
+        module_imports = "\n".join(module_dict["session_imports"])
+        artifact_functions = "\n\n".join(module_dict["artifact_functions"])
```

Comment:
purposefully two newlines here?

Reply:
Yes, initially, but we should no longer need it as formatting is taken care of by `prettify()` utility. So, I will make it a single `\n`